llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -55,15 +56,91 @@ void InterBlockScheduling::markEpilogueBlocks() {
   }
 }
 
+std::map<MachineBasicBlock *, LivePhysRegs> calculateLiveOuts(MachineFunction *MF) {
+  std::queue<MachineBasicBlock *> WorkList;
+  std::set<MachineBasicBlock *> WorkListSet;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveIns;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveOuts;
+  const TargetRegisterInfo *TRI = MF->getSubtarget().getRegisterInfo();
+  for (MachineBasicBlock *MBB : post_order(MF)) {
+    LiveIns[MBB].init(*TRI);
+    LiveOuts[MBB].init(*TRI);
+    WorkListSet.insert(MBB);
+    WorkList.push(MBB);
+  }
+  while (!WorkList.empty()) {
+    MachineBasicBlock *MBB = WorkList.front();
+    WorkList.pop();
+    WorkListSet.erase(MBB);
+    LivePhysRegs CurrentLive;
+    CurrentLive.init(*TRI);
+
+    // Calculates LiveOuts by iterating over each successor of the current MBB
+    // and adding each successor's LiveIns to the current MBB's LiveOuts.
+    // This ensures LiveOuts is the union of LiveIns of all successor MBBs.
+    for (MachineBasicBlock *Succ : MBB->successors()) {
+      for (MCPhysReg Reg : LiveIns[Succ]) {
+        LiveOuts[MBB].addReg(Reg);
+        CurrentLive.addReg(Reg);
+      }
+    }
+    // Calculates CurrentLive by bottom-up traversal the MBB
+    for (MachineInstr &MI : llvm::reverse(*MBB)) {
+      CurrentLive.stepBackward(MI);
+    }
+    // Checks for convergence of LiveIns of MBB
+    bool Change = false;
+    int LiveInCount = 0;
+    for (MCPhysReg Reg : LiveIns[MBB]) {
+      if (!CurrentLive.contains(Reg)) {
+        Change = true;
+        break;
+      }
+      LiveInCount++;
+    }
+    LiveIns[MBB].init(*TRI);
+    // Updates LiveIns of MBB with CurrentLive
+    for (MCPhysReg Reg : CurrentLive) {
+      LiveInCount--;
+      LiveIns[MBB].addReg(Reg);
+    }
+    if (LiveInCount != 0) {
```

Comment:
I don't understand the point of `LiveInCount`. Why do we need to track this? As far as I understand, we just need to re-run the analysis for MBB's predecessors if the liveins of MBB change. That is, if `CurrentLive` differs from `LiveIns[MBB]`. Is there something I'm missing?

Reply:
We need to check for convergence of LiveIns (CurrentLive and LiveIns[MBB]) for termination of loop, as there no '==' operator for LivePhysRegs. For comparison I used LiveInCount. LiveIns[MBB] is old liveins and CurrentLive is new liveins.